src/EfuncMsg.chpl

```diff
@@ -1375,4 +1500,58 @@ module EfuncMsg
       }
       return C;
     }
+//}
+//}
+    proc putmask_helper (mask : [?D1] bool, ref A : [D1] ?ta, Values : [?D2] ?tv) {
+
+//  Note, added casting to the output, to accomodate the various combinations of
+//  types that are allowed by np.putmask.  Invalid pairings of types ta, tv must be 
+//  caught and flagged by the calling function.  This proc does no type checking.
+
+      if A.size == Values.size { // then their distributions match
+
+	forall element in A.domain {
+		if mask[element] then A[element] = Values[element]:ta ;
+	}	
+
+      } else if A.size < Values.size { // then prune Values so the distributions match
+
+	var Pruned_Values = Values[0..A.size-1] ;
+	forall element in A.domain {
+		if mask[element] then A[element] = Pruned_Values[element]:ta ;
+	}	
+
+//  For the remaining cases, we know Values.size < A.size
+
+      } else {
+	var A_subset_size = (A.size/numLocales):int ;
+	if A.size%numLocales == 0 then A_subset_size += 1 ;
+
+//  If Values.size < A_subset_size, then we may as well make a copy of Values on each locale
+
+	if Values.size < A_subset_size {
+		coforall loc in Locales do on loc {
+			var local_Values = Values ;
+			forall element in A.localSubdomain() {
+				if mask[element] then A[element] = (local_Values[element%local_Values.size]):ta ;
+			}
+		}
+
+//  Finally, we have the case where Values.size > A_subset_size, which means Values is distributed
+//  but probably not to our liking.  Use an aggregator to get the right values in the right places.
+
+	} else {
+		coforall loc in Locales do on loc {
+			var local_Values : [0..A.localSubdomain().size-1] tv;
+			forall (element, item) in zip (A.localSubdomain(),local_Values) with (var agg = newSrcAggregator(tv)) {
+				agg.copy(item,Values[element%Values.size]) ;
+			}
+			forall (element, item) in zip (A.localSubdomain(),local_Values) {
```

Comment:
I'm not sure that will have the right affect, but I'll try it.

Reply:
It works.  It even makes local_Values unnecessary.  Cool.

Reply:
Okay, no, not as simple as I'd hoped.  This seems to introduce a bug.  The aggregate test succeeds with two forall loops, and fails with one.  This might be a good candidate for some team analysis.

Reply:
I've saved a jupyter notebook showing the error.

Reply:
Looks like the error may have something to do with changing the type of the aggregator.